frame/uniswap-v2/src/lib.rs

```diff
@@ -0,0 +1,857 @@
+			admin_fees: Vec<T::Balance>,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {}
+
+	impl<T: Config> CurveAmm for Pallet<T> {
+		type AssetId = T::AssetId;
+		type Balance = T::Balance;
+		type AccountId = T::AccountId;
+		type PoolId = T::PoolId;
+		type PoolTokenIndex = T::PoolTokenIndex;
+
+		fn pool_count() -> T::PoolId {
+			PoolCount::<T>::get()
+		}
+
+		fn add_liquidity(
+			who: &Self::AccountId,
+			pool_id: T::PoolId,
+			amounts: Vec<Self::Balance>,
+			min_mint_amount: Self::Balance,
+		) -> Result<(), DispatchError> {
+			let zero = Self::Balance::zero();
+			ensure!(amounts.len() == 2, Error::<T>::IndexOutOfRange);
+			ensure!(amounts.iter().all(|&x| x >= zero), Error::<T>::WrongAssetAmount);
+
+			let pool_lp_asset =
+				PoolLPAsset::<T>::get(pool_id).ok_or(Error::<T>::InconsistentStorage)?;
+			let assets = PoolAssets::<T>::get(pool_id).ok_or(Error::<T>::InconsistentStorage)?;
+			let mut balances = Vec::new();
+			for asset_id in &assets {
+				balances.push(PoolAssetBalance::<T>::get(pool_id, asset_id));
+			}
+
+			let n_coins = assets.len();
+
+			ensure!(n_coins == balances.len(), Error::<T>::InconsistentStorage);
+
+			ensure!(n_coins == amounts.len(), Error::<T>::IndexOutOfRange);
+
+			let old_balances: Vec<FixedU128> = balances
+				.iter()
+				.map(|b| {
+					let b_u: u128 = (*b).into();
+					FixedU128::saturating_from_integer(b_u)
+				})
+				.collect();
+			let old_balance_a = old_balances[0];
+			let old_balance_b = old_balances[1];
+			let amount_a_u: u128 = amounts[0].into();
+			let amount_b_u: u128 = amounts[1].into();
+			let amount_a_desired = FixedU128::saturating_from_integer(amount_a_u);
+			let amount_b_desired = FixedU128::saturating_from_integer(amount_b_u);
+			let amount_a;
+			let amount_b;
+
+			if old_balance_a == FixedU128::zero() && old_balance_b == FixedU128::zero() {
+				amount_a = amount_a_desired;
+				amount_b = amount_b_desired;
+			} else {
+				let amount_b_optimal = Self::quote(amount_a_desired, old_balance_a, old_balance_b)
+					.ok_or(Error::<T>::Math)?;
+
+				if amount_b_optimal <= amount_b_desired {
+					amount_a = amount_a_desired;
+					amount_b = amount_b_optimal;
+				} else {
+					let amount_a_optimal =
+						Self::quote(amount_b_desired, old_balance_b, old_balance_a)
+							.ok_or(Error::<T>::Math)?;
+					assert!(amount_a_optimal <= amount_a_desired);
+					amount_a = amount_a_optimal;
+					amount_b = amount_b_desired;
+				}
+			}
+
+			let token_supply = T::LpToken::total_issuance(pool_lp_asset);
+			let token_supply_u: u128 = token_supply.into();
+			let token_supply_f = FixedU128::saturating_from_integer(token_supply_u);
+			let mint_amount;
+			if token_supply_u == 0 {
+				let mint_amount_f = amount_a.checked_mul(&amount_b).ok_or(Error::<T>::Math)?;
+				let mint_amount_b: T::Balance =
+					mint_amount_f.checked_mul_int(1u64).ok_or(Error::<T>::Math)?.into();
+				mint_amount = mint_amount_b.integer_sqrt();
+			} else {
+				let mint_a = amount_a
```

Comment:
it is pure math, can we move it into math.rs?

Reply:
Is it okay I do a separate PR for that?